### PR TITLE
Experimental support and make pop public 

### PR DESCRIPTION
**Pull request by bgavrilMS**

### CODE REVIEW

**Comment by henrik-me** [comment]
was immediately thinking about the need to enable/disable specific experimental features... but really this is likely what we will ever need :)

File: src/client/Microsoft.Identity.Client/MsalErrorMessage.cs [hunk 1]
```diff
@@ -286,23 +286,32 @@ public static string RedirectUriMismatch(string expectedUri, string actualUri)
             " See https://aka.ms/msal-net-invalid-client for details. ";
         public const string SSHCertUsedAsHttpHeader = "MSAL was configured to request SSH certificates from AAD, and these cannot be used as an HTTP authentication header. Developers are responsible for transporting the SSH certificates to the target machines.";
 
+        public static string ExperimentalFeature(string methodName)
+        {
+            return string.Format(
+                CultureInfo.InvariantCulture,
+                "The API {0} is marked as experimental. It may change without incrementing the major version of the library. Call .WithExperimentalFeatues() when creating the public / confidential client to bypass this.",
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/MsalErrorMessage.cs]: sp.....WithExperimentalFeatures()

Comment (henrik-me) [reply src/client/Microsoft.Identity.Client/MsalErrorMessage.cs]: should probably have a disclaimer saying that this one might not do anything and should not be used in production?

File: src/client/Microsoft.Identity.Client/AppConfig/IAppConfig.cs [hunk 2]
```diff
@@ -84,6 +84,12 @@ public interface IAppConfig
         /// </summary>
         ITelemetryConfig TelemetryConfig { get; }
 
+        /// <summary>
+        /// Allows usage of features that are documented as experimental and would otherwise throw a specific exception
+        /// </summary>
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/AppConfig/IAppConfig.cs]: >summary [](start = 14, length = 7)

add a remark saying: Do not use in production

**Comment by bgavrilMS** [comment]
Also, @henrik-me and @jmprieur - I don't think PoP made it to prod yet, but I doubt anybody will pick it up without handholding, so should be ok...

File: src/client/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs [hunk 3]
```diff
@@ -295,6 +295,24 @@ public T WithExtraQueryParameters(string extraQueryParameters)
...
+
+        /// <summary>
+        /// Allows usage of experimental features and APIs. If this flag is not set, experimental features 
+        /// will throw an exception. 
+        /// </summary>
+        /// <remarks>
+        /// Changes in the public API of experimental features will not result in an increment of the major version of this library.
+        /// The same level of support is given to experimental features as to all other features.
+        /// Affected APIs are clearly documented as being experimental. 
+        /// </remarks>
```
Comment (henrik-me) [inline src/client/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs]: ></remarks> [](start = 12, length = 10)

Use of experimental features in production is not recommended and is subject to be removed between builds

**Comment by henrik-me** [comment]
>         /// <item>This is an experimental API. The method signature may change in the future without involving a major version upgrade.</item>

Should we add a "deprecated" attribute for these features ensuring that people see this in their builds?

---
Refers to: src/client/Microsoft.Identity.Client/ApiConfig/AbstractPublicClientAcquireTokenParameterBuilder.cs:37 in 70ac55d. [](commit_id = 70ac55dd409fa672a8dd641c4028b392ade3c5c5, deletion_comment = False)

File: src/client/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs (hunk 3 above)
Comment (jmprieur) [reply src/client/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs]: Unless you are directed to, use of experimental features ... ?

File: src/client/Microsoft.Identity.Client/MsalError.cs [hunk 4]
```diff
@@ -715,6 +715,12 @@ public static class MsalError
         /// </summary>
         public const string TokenTypeMismatch = "token_type_mismatch";
 
+        /// <summary>
+        /// <para>What happens?</para>You are trying to use a feature which was marked as experimental
+        /// <para>Mitigation</para>When creating the PublicClientApplication or ConfidentialClientApplication, use .WithExperimentalFeatures() option.
```
Comment (jmprieur) [inline src/client/Microsoft.Identity.Client/MsalError.cs]: See aka.ms/msal-net-experimental-features ?

File: src/client/Microsoft.Identity.Client/MsalErrorMessage.cs (hunk 1 above)
Comment (jmprieur) [inline src/client/Microsoft.Identity.Client/MsalErrorMessage.cs]: See https://aka.ms/msal-net-experimental-features

Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/MsalErrorMessage.cs]: @henrik-me - some of our "experimental" stuff, like opening the system browser using Edge, is already used in production. I'll add the disclaimer and go through our existing stuff to see what we can remove.

File: src/client/Microsoft.Identity.Client/MsalError.cs (hunk 4 above)
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/MsalError.cs]: Is there anything else that we need to add in a wiki page? I feel that the 3 lines of comments pretty much explain the whole thing.

Comment (jmprieur) [reply src/client/Microsoft.Identity.Client/MsalError.cs]: yes, as discussed, a list of the current experimental features.

**Pull request merged**